csp/impl/types/typing_utils.py

```diff
@@ -54,7 +57,165 @@ def get_orig_base(cls, typ):
             return numpy.ndarray
         return res
 
+    @classmethod
+    def pretty_typename(cls, typ):
+        if cls.is_generic_container(typ):
+            return str(typ)
+        elif cls.is_forward_ref(typ):
+            return cls.pretty_typename(typ.__forward_arg__)
+        elif isinstance(typ, type):
+            return typ.__name__
+        else:
+            return str(typ)
+
 
-# Current typing utilities were
-# stabilized as of python 3.7
 CspTypingUtils = CspTypingUtils37
+
+if sys.version_info >= (3, 9):
+
+    class CspTypingUtils39(CspTypingUtils37):
+        # To support PEP 585
+        _GENERIC_ALIASES = (typing._GenericAlias, typing.GenericAlias)
+
+    CspTypingUtils = CspTypingUtils39
+
+if sys.version_info >= (3, 10):
+
+    class CspTypingUtils310(CspTypingUtils39):
+        # To support PEP 604
+        @classmethod
+        def is_union_type(cls, typ):
+            return (isinstance(typ, typing._GenericAlias) and typ.__origin__ is typing.Union) or isinstance(
+                typ, types.UnionType
+            )
+
+    CspTypingUtils = CspTypingUtils310
```

Comment:
this will get defined twice , dont think that was intentional ( ie py310 will also pass the 3,9 check above )

Reply:
It was, and it actually gets defined three times because it starts out as `CspTypingUtils = CspTypingUtils37`. The tricky part is that `CspTypingUtils310` depends on `CspTypingUtils39`, which depends on `CspTypingUtils37`, but they can only be defined in those versions of python. So either CspTypingUtils gets re-defined as necessary, or there needs to be another if-else block (in reverse order) defining CspTypingUtils only once. It didn't look cleaner when I did it that way.

Reply:
I see.  I guess it kind of makes sense, subsequent versions have changes and they accumulate up